### ISSUE TITLE
RT_SPLIT

### ISSUE THREAD

**Issue by bencannell** [tkoziara/solfec#67]
As discussed, see the attached file defining the documentation/assumptions of the proposed RT_SPLIT function to be included in Solfec. The overall aim of this function is to provide the capability to split a single BODY into two BODY objects. The function will modify the BODY being split, and produce one new BODY. It should be possible to call the RT_SPLIT function during the simulation (e.g. using the CALLBACK utility).

[RT_SPLIT.txt](https://github.com/tkoziara/solfec/files/1063957/RT_SPLIT.txt)

**Comment by tkoziara** [comment]
Ben,
Questions:
1. What should be the behaviour of lst1 in case of a split that only opens a partial crack and does not create two fragments? (I am assuming lst2 to be *None* in this case) We are going to have more nodes than originally because of node duplications on the surface that is opening up.
2. Should the procedure abort with an error incase more than two fragments are created?
3. The procedure specification needs to be modified so that changes to body labels are clearly defined. The reason for this is that for this procedure to work in parallel - the user will always need to use BYLABEL to initially obtain the BODY object - just before calling RT_SPLIT; this is because in parallel bodies keep migrating between MPI ranks and a previously stored BODY object is cannot be depended on; surely you could store individual BODY objects for particular bodies and use HERE to check whether it makes sense to call RT_SPLIT - but in this case you would need to store the exact same set of separate BODY objects and then check them all every time, in your callback; either case should work but still a convention for modifying labels is needed for completeness - would you like to propose one?

I am mentioning @stevebrasier so that he also gets notified:)
Tomek

**Comment by bencannell** [comment]
Hi Tomek,
In response to your questions:

1. Good point! The most robust solution I can think of would be to change lst1, lst2 to dict1, dict2, which are python dictionaries where key:=new_node_id, value:=original_node_id (presumably sometimes new_node_id=original_node_id, this is OK). This would also work for cases where two fragments are created. For the case you describe (partial crack), where the number of nodes increases, the original_node_id associated with the new_node_id, should be the original node which is coincident with the new node (and dict2=_None_).
2. Yes.
3. We don't use BYLABEL to identify objects, we required more detailed information about the component that each BODY represents so we wrap all of the Solfec BODY objects with python objects. We purely write to BODY.label to conveniently identify a BODY in Viewer. As you suggest, we intended on using the HERE command to check that the specific BODY object existed on the MPI rank. For completeness, could the default setting be for the newly created body to not have a label (as is currently described in the BODY documentation)? All we require is that we can write to BODY.label when we call the CALLBACK command.

Thanks,
Ben

**Comment by tkoziara** [comment]
OK, thank you.

Re 1: In fact there is no need for a dictionary - it is enough to have ``lst1`` as a longer list referencing original mesh indices multiple times; somehow this possibility seemed obscured prior to your comment:)

Re 3: I added optional ``label1`` and ``label2`` arguments that are empty by default.

**Comment by tkoziara** [comment]
Hi @bencannell and @stevebrasier :)

FYI: on the way to this functionality I extended the [MESH_SPLIT](http://parmes.org/solfec/user/utilities.html#mesh-split) command so that it can now fragment meshes by using face lists. Here is a [test example](https://github.com/tkoziara/solfec/blob/master/inp/devel/mesh_split.py).

Looking further into this - there are subtleties of having a command that splits an existing body in real time from a callback - and returns the other body back to Python interpreter:
* in parallel this will happen only on one MPI rank, where the split body resides;
* RT_SPLIT may return *None* as a result of a single split - the other body is not produced; or it may return *None* as a result of being called on an MPI rank where the split body is not present; to work around this you will need to use HERE(...) to check whether a body is present - and only then call RT_SPLIT;
* when a second body is returned after RT_SPLIT has been called - in parallel this will happen only on one MPI rank and consequently you will not be able to refer to this body on other ranks; labelling of split bodies will be optionally possible and this way, by using BYLABEL *at subsequent time steps* you will be able to access such body on other MPI ranks (when it migrates away from the rank where the split happened);

I am sharing those thoughts for our record and also - you may want to comment and further guide development of this functionality based on your needs:)

**Comment by tkoziara** [comment]
@bencannell , @stevebrasier 

The initial version of [RT_SPLIT](http://parmes.org/solfec/user/utilities.html#rt-split) is now enabled and tested in serial mode; It does not yet work in parallel; A test example is [here](https://github.com/tkoziara/solfec/blob/master/inp/devel/rt_split.py);

One base **deviation** from your assumptions is that the ID of the original body *does change* when two fragments are created. This is necessary for the Solfec file writing/reading to work correctly when body fragmentation happens. Essentially we give a new ID to every new body until all possible IDs run out (few billion) and this way can have unique identification of all bodies and their fragments during entire time history. When reading, Solfec stores all those bodies and is able to jump one time instant to another without loosing track of the bodies. I hope that you will be able to work around this aspect of the implementation.

Please, feel free to review the current version and provide feedback. In the meantime I will look into enabling it in parallel and perhaps relaxing some of the current limitations (e.g. constraints could be remapped to fragments rather than removed as it is currently).

**Comment by tkoziara** [comment]
@bencannell , @stevebrasier 

Just letting know that both serial and parallel versions are now ready for testing; An example can be found in: [inp/devel/rt_split.py](https://github.com/tkoziara/solfec/blob/master/inp/devel/rt_split.py)

**Comment by tkoziara** [comment]
@bencannell , @stevebrasier 

Added a clarification that the ID of body1 changes every time a split happens; [See point 3](http://parmes.org/solfec/user/fragmentation.html#solfec-user-fragmentation).
This is the easiest way to go for now because of how fragmentation is handled in Solfec; Linking back to the explanation provided above;

**Comment by bencannell** [comment]
Hi,

I've just tested RT_SPLIT() on a couple of test meshes and while it appears to correctly split hexahedral meshes it does not correctly split tetrahedral meshes (the mesh has some evidence of new faces being created, but separate bodies are not returned). I'll try and get an example case together on Monday to send over.

I had similar issues while developing our own mesh splitting function. These were caused by me not handling elements with nodes, but not faces, on the split plane. I ended up having to check the connectivity of these elements (i.e. shared faces) with elements which had faces on the split plane. The "side" of the split plane which these elements ended up on was determined by whether they solely had connectivity with elements on the "positive" or "negative" side of the plane.

**Comment by bencannell** [comment]
FYI - the above tests were on changeset cc02419. I'll update our version of solfec on Monday and re-test...

**Comment by bencannell** [comment]
Hi.

I've tested the code using solfec changeset f7041ed and the issue with splitting tet meshes still exists.

I'll email you a test case demonstrating the issue.

**Comment by tkoziara** [comment]
Just letting know that I am working on this also; Generated a tet mesh for one of the pipes in the example and run into issues as well; Your example will additionally help - thank you;

**Comment by bencannell** [comment]
There seems to be a bug in RT_SPLIT when I supply optional *kwargs.

For example, the following works correctly:

RT_SPLIT(body, surf, 1)

Whereas:

RT_SPLIT(body, surf, sid1=1)

Returns a “TypeError: an integer is required”.

**Comment by tkoziara** [comment]
I spotted this and fixed in 5d4a433; Perhaps you've come across this prior to that commit?

**Comment by tkoziara** [comment]
I spotted this and fixed in 5d4a433; Perhaps you've come across this prior to that commit?

**Comment by bencannell** [comment]
@tkoziara 

How would I go about selecting a point on a split body for extracting displacement results if the body is not in the solfec.bodies list at t=0s? Do I need to pass HISTORY or DISPLACEMENT a "point" on the body at t=[time of split]?

**Comment by tkoziara** [comment]
That's a good point, @bencannell :)
I think we may need to implement a modification of the HISTORY command to handle that; It could perhaps accept a string label of the body and then retrieve the history only when the body existed;

**Comment by bencannell** [comment]
Yes, that would work. It would be most convenient for us if updated HISTORY command could handle the body associated with a label changing throughout the history i.e. so we can easily refer to "the same" component even though the body.id changes following fragmentation. Although, we can work around this if you think it's too much of a potential error trap...?

@stevebrasier for info

**Comment by bencannell** [comment]
It would also be most convenient if we could still access results by providing the "point" on the body at t=0s (even if the body doesn't exist yet)?

**Comment by tkoziara** [comment]
Yes, perhaps this would be possible as well - to simply keep referring to the initial body - while the actual point on the body would migrate from the initial body to a suitable fragment.

**Comment by tkoziara** [comment]
Let's create another issue where all those ideas are summarised; Something like: "HISTORY refinements for fragmentation"; Would you?:)

**Comment by tkoziara** [comment]
OK, just created #70 ; Please add any comments regarding that there;

**Comment by tkoziara** [comment]
@bencannell @stevebrasier RT_SPLIT now works correctly with half-split surfaces; Here are pics from your two tests:
<img width="216" alt="rt_split_half1" src="https://user-images.githubusercontent.com/2198068/31223572-0d0eea8c-a9cb-11e7-8726-e806f4c0d896.png"><img width="213" alt="rt_split_half2" src="https://user-images.githubusercontent.com/2198068/31223588-1b712d92-a9cb-11e7-919a-3b6d164d2bfb.png">

**Comment by bencannell** [comment]
@tkoziara great, I let you know if I have any issues with our tests :)

**Comment by bencannell** [comment]
Hi @tkoziara, yes :)
